pkgs/sketch_pad/lib/main.dart

```diff
@@ -779,6 +801,77 @@ class ListSamplesWidget extends StatelessWidget {
   }
 }
 
+class SelectLocaleWidget extends StatelessWidget {
+  final void Function(Locale l) handleLocaleChanged;
+  const SelectLocaleWidget({super.key, required this.handleLocaleChanged});
+
+  @override
+  Widget build(BuildContext context) {
+    final appServices = Provider.of<AppServices>(context);
+
+    return ValueListenableBuilder<Channel>(
+      valueListenable: appServices.channel,
+      builder: (context, Channel value, _) {
+        return SizedBox(
+          height: toolbarItemHeight,
+          child: TextButton.icon(
+            icon: const Icon(Icons.language, size: smallIconSize),
+            label: Text(getDisplayName(Localizations.localeOf(context))),
+            onPressed: () async {
+              final selection = await _showMenu(
+                context,
+                calculatePopupMenuPosition(context, growUpwards: true),
+                value,
+              );
+              if (selection != null && context.mounted) {
+                _handleSelection(context, selection);
+              }
+            },
+          ),
+        );
+      },
+    );
+  }
+
+  Future<Locale?> _showMenu(
+      BuildContext context, RelativeRect position, Channel current) {
+    const itemHeight = 46.0;
+
+    final menuItems = [
+      for (final locale in context
+          .findAncestorWidgetOfExactType<MaterialApp>()!
+          .supportedLocales)
+        PopupMenuItem<Locale>(
+          value: locale,
+          child: PointerInterceptor(
+            child: ListTile(
+              title: Text(getDisplayName(locale)),
+            ),
+          ),
+        )
+    ];
+
+    return showMenu<Locale>(
+      context: context,
+      position: position.shift(Offset(0, -1 * menuItems.length * itemHeight)),
+      items: menuItems,
+    );
+  }
+
+  String getDisplayName(Locale locale) => intl.Intl()
+      .displayNames()
+      .ofLanguage(intl.Locale.parse(locale.toLanguageTag()));
+
+  void _handleSelection(BuildContext context, Locale locale) async {
+    final appServices = Provider.of<AppServices>(context, listen: false);
+    handleLocaleChanged(locale);
+    appServices.appModel.editorStatus.showToast(
```

Comment:
For some reason, this message is still shown with the old locale. I don't know why - does `getState` take a while to propagate?

Reply:
It looks like `handleLocaleChanged` calls `setState`? IIRC those closures (to setState) won't be called immediately. I forget the exact lifecycle but they'll be queued up and dispatched later.

If this is the only call to handleLocaleChanged you might just put the setstate call in-line here:

```
setState(() {
  this.locale = locale;
});
```

And in terms of getting the correct locale w/ the notification toast, you could just try not having a toast? Having the rest of the UI text update may be enough of a notification to users that their action took effect.